README.md
[![Workflow Status](https://github.com/tehzz/vpk0/workflows/main/badge.svg)](https://github.com/tehzz/vpk0/actions?query=workflow%3A%22main%22)

# vpk0

A Rust library for handling Nintendo's N64-era `vpk0` data compression

`vpk0` is data compression scheme built on LZSS with Huffman coding for the
dictionary offsets and match lengths.
It was used at HAL Laboratory for three of their N64 games and later,
for encoding the data on GBA E-Reader cards.

This crate provides decoding and encoding for `vpk0` files.
As it is an old scheme, the decoding and encoding is designed to match what
Nintendo did in the 1990s; the crate does not focus on compression ratio or speed.

### Usage
The [`decode()`] and [`encode()`] functions provide quick ways to deal with
`vpk0` data.

```rust
use vpk0::{encode, decode};
use std::io::Cursor;

let raw = Cursor::new(data);
let compressed = encode(raw).unwrap();
let decompressed = decode(Cursor::new(&compressed)).unwrap();
assert_eq!(&data, &decompressed);
```

For more control, you can use [`Decoder`] or [`Encoder`]:

```rust
use vpk0::Encoder;

Encoder::for_bytes(b"ababacdcdeaba")
    .two_sample()
    .encode_to_writer(std::io::stdout())
    .unwrap();
```
--------------------------------------------------------------------------------
### `vpk0` Background
The `vpk0` format—named for the magic bytes—is thought to have been developed
at [HAL Laboratories] in the late 1990s. Three of their N64 games use the compression scheme:
* Super Smash Bros.
* Pokémon Snap
* Shigesato Itoi's No. 1 Bass Fishing: Definitive Edition

The format next appeared in the mid-2000s as the compression used in the [Nintendo e-Reader]
for the GBA. This is where the format first received attention from the internet at large.
Tim Schuerewegen’s [nevpk] and Caitsith’s [NVPK Tool and NEDEC Make] were
open source implementations of `vpk0` that came from reverse engineering the e-Reader.

This crate extends on their work to provide matching compression for HAL’s N64 titles.

### Format Overview
`vpk0` is based on two fundamental encoding algorithms: LSZZ and Huffman Coding.
The two techniques were bouncing around [the Japanese BSSes since the late 80s],
and together they comprise the backbone of many modern day encoding schemes like [Deflate].

The `vpk0` format is comparatively simpler: it is a variable length LZSS.
The input data is compressed by a standard LZSS implementation.
But instead of having fixed bit sizes for the dictionary offset and length, the sizes are variable.
The variable bit sizes are then encoded as a Huffman code, with the necessary
[Huffman tree](format::TreeInfo) prepended to the encoded data.

For more info, see [the documentation for the format module](crate::format)

### Implementation Details
This implementation is designed to be a byte-perfect match of the encoder used
for *Super Smash Bros.* As of March 2021, the LZSS encoder is byte-matching,
but the Huffman compression is not.

The matching LZSS encoding scheme is: after a found match, look ahead at the next byte
to see if there is a longer match. Continue checking the next byte until
a smaller or no match is found.

The encoder in this crate checks at most the next ten bytes,
as that was the maximum number necessary to match all 500 `vpk0` encoded files in *SSB64*.
In the future, this parameter may become another option for [`Encoder`].

### Advanced Usages
#### Getting info from a `vpk0` file
```rust
use vpk0::vpk_info;
let (header, trees) = vpk_info(vpkfile).unwrap();
println!("Original size: {} bytes", header.size);
println!("VPK encoded with method {}", header.method);
println!("Offsets: {} || Lengths: {}", trees.offsets, trees.lengths);
```

#### Encode like a standard LZSS
```rust
use vpk0::{Encoder, LzssSettings};
// use fixed length compression by setting the offset to 10 and the length to 6.
let compressed = Encoder::for_bytes(b"I am Sam. Sam I am.")
    .one_sample()
    .with_lzss_settings(LzssSettings::new(10, 6, 2))
    .with_offsets("10")
    .with_lengths("6")
    .encode_to_vec();
```

[HAL Laboratories]: https://www.hallab.co.jp/eng/
[Nintendo e-Reader]: https://en.m.wikipedia.org/wiki/Nintendo_e-Reader
[nevpk]: http://users.skynet.be/firefly/gba/e-reader/tools/index.htm
[NVPK Tool and NEDEC Make]: https://caitsith2.com/ereader/devtools.htm
[the Japanese BSSes since the late 80s]: https://web.archive.org/web/20160110174426/https://oku.edu.mie-u.ac.jp/~okumura/compression/history.html
[Deflate]: https://en.m.wikipedia.org/wiki/Deflate

License: MIT


CHANGELOG.md
# Changelog

## [0.8.2] 2022-01-15
### Added
* Two helper functions for handling `&[u8]` data: `encode_bytes` and `decode_bytes`

### Changed

### Fixed
* Encoding data that happens to be uncompressible (e.g. `"123456789"`) will not crash

## [0.8.1] 2021-03-11
Initial release. Supports decoding and encoding vpk0 data.


tests/tests.rs
use std::io::Cursor;
use vpk0::LzssBackend::{self, *};

const LOGO: &[u8] = include_bytes!("small-logo.png");
const BACKENDS: &[LzssBackend] = &[Brute, Kmp, KmpAhead];
const VPK_METHOD0: &[u8] = include_bytes!("method0.vpk0");
const RAW_METHOD0: &[u8] = include_bytes!("method0-orig.bin");
const VPK_METHOD1: &[u8] = include_bytes!("method1.vpk0");
const RAW_METHOD1: &[u8] = include_bytes!("method1-orig.bin");
const NOREPT_MESG: &[u8] = b"abcdefgh12345";

#[test]
fn decode_method0() {
    let decoded = vpk0::Decoder::for_bytes(VPK_METHOD0)
        .decode()
        .expect("working decode");

    assert_eq!(decoded, RAW_METHOD0, "decoding method 0");
}

#[test]
fn decode_method0_bytes() {
    let decoded = vpk0::decode_bytes(VPK_METHOD0).expect("working decode");

    assert_eq!(decoded, RAW_METHOD0, "decoding method 0");
}

#[test]
fn encode_method0() {
    for &backend in BACKENDS {
        vpk0::Encoder::for_bytes(LOGO)
            .one_sample()
            .lzss_backend(backend)
            .encode_to_vec()
            .expect(&format!("valid encode for {:?}", backend));
    }
}

#[test]
fn encode_method0_bytes() {
    vpk0::encode_bytes(LOGO).expect("encode bytes helper function works");
}

#[test]
fn encode_method0_no_matching() {
    for &backend in BACKENDS {
        let compressed = vpk0::Encoder::for_bytes(NOREPT_MESG)
            .one_sample()
            .lzss_backend(backend)
            .encode_to_vec()
            .expect(&format!("encoding no match data for {:?}", backend));

        let mut decoder = vpk0::Decoder::for_bytes(&compressed);

        let trees = decoder.trees().expect("valid zero length VPK trees");
        assert_eq!(trees.lengths, "()", "Expected empty length tree");
        assert_eq!(trees.offsets, "()", "Expected empty offset tree");

        let decompressed = decoder.decode().expect("decoding of empty tree vpk0 file");

        assert_eq!(
            NOREPT_MESG, decompressed,
            "Expected encoded-decoded data to match original data"
        );
    }
}

#[test]
fn match_method0() {
    let (_header, trees) = vpk0::vpk_info(Cursor::new(VPK_METHOD0)).unwrap();

    let compressed = vpk0::Encoder::for_bytes(RAW_METHOD0)
        .one_sample()
        .lzss_backend(Brute)
        .with_lengths(&trees.lengths)
        .with_offsets(&trees.offsets)
        .encode_to_vec()
        .unwrap();

    assert_eq!(compressed, VPK_METHOD0);
}

#[test]
fn decode_method1() {
    let mut reader = Cursor::new(VPK_METHOD1);
    let decoded = vpk0::decode(&mut reader).unwrap();

    assert_eq!(decoded, RAW_METHOD1, "error method 1");
}

#[test]
fn decode_method1_bytes() {
    let decoded = vpk0::decode_bytes(VPK_METHOD1).unwrap();

    assert_eq!(decoded, RAW_METHOD1, "error method 1");
}

#[test]
fn encode_method1() {
    for &backend in BACKENDS {
        vpk0::Encoder::for_bytes(LOGO)
            .two_sample()
            .lzss_backend(backend)
            .encode_to_vec()
            .expect(&format!("valid encode for {:?}", backend));
    }
}

#[test]
fn encode_method1_no_matching() {
    for &backend in BACKENDS {
        let compressed = vpk0::Encoder::for_bytes(NOREPT_MESG)
            .two_sample()
            .lzss_backend(backend)
            .encode_to_vec()
            .expect(&format!("encoding no match data for {:?}", backend));

        let mut decoder = vpk0::Decoder::for_bytes(&compressed);

        let trees = decoder.trees().expect("valid zero length VPK trees");
        assert_eq!(trees.lengths, "()", "Expected empty length tree");
        assert_eq!(trees.offsets, "()", "Expected empty offset tree");

        let decompressed = decoder.decode().expect("decoding of empty tree vpk0 file");

        assert_eq!(
            NOREPT_MESG, decompressed,
            "Expected encoded-decoded data to match original data"
        );
    }
}

#[test]
fn match_method1() {
    let (_header, trees) = vpk0::vpk_info(Cursor::new(VPK_METHOD1)).unwrap();

    let compressed = vpk0::Encoder::for_bytes(RAW_METHOD1)
        .two_sample()
        .lzss_backend(Brute)
        .with_lengths(&trees.lengths)
        .with_offsets(&trees.offsets)
        .encode_to_vec()
        .unwrap();

    assert_eq!(compressed, VPK_METHOD1);
}

#[test]
fn decode_bad_file() {
    let bad_file = include_bytes!("bad-file.vpk0");
    let mut bf_r = Cursor::new(bad_file.as_ref());

    match vpk0::decode(&mut bf_r) {
        Ok(result) => {
            let err = "Expected error when decoding bad file";
            eprintln!("{}", err);
            eprintln!("{:?}", result);
            panic!("Expected error when decoding bad file");
        }
        Err(err) => {
            eprintln!("{}", err);
            assert!(true)
        }
    };
}


src/decode.rs
use crate::errors::VpkError;
use crate::format::{VpkHeader, VpkMethod, VpkTree};
use bitstream_io::{BigEndian, BitReader};
use std::{
    collections::BTreeMap,
    fs::File,
    io::{BufReader, Cursor, Read, Write},
    path::Path,
};

type LogWtr<'a> = &'a mut dyn Write;
// [offset, length]
type RawTrees = [VpkTree; 2];

/// Textual representations of the offsets and lengths Huffman trees in a `vpk0` file
///
/// The trees are comprised of decimal numbers—the leafs—separated by commas and parentheses—the nodes.
/// The trees also follow the typical Huffman Tree convention of `0` for left nodes
/// and `1` for right nodes. So, if you have `((4, 1), (8, (15, 10))`,
/// `4` would have the Huffman code `00` and `15` would have the Huffman code `110`.
///
/// You can get a `TreeInfo` by using [`vpk_info`] or [`Decoder::trees`].
/// The `String`s can then be used by [`Encoder::with_offsets`] (and related functions)
/// to set the offsets and lengths tree in a new encode.
///
/// [`Encoder::with_offsets`]: crate::Encoder::with_offsets()
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TreeInfo {
    pub offsets: String,
    pub lengths: String,
}

impl From<&RawTrees> for TreeInfo {
    fn from([offsets, lengths]: &RawTrees) -> Self {
        Self {
            offsets: offsets.to_string(),
            lengths: lengths.to_string(),
        }
    }
}

/// Specify the decoding settings, such as logging, input, and output.
///
/// To create a new `Decoder`, use [`for_reader()`], [`for_bytes()`], or
/// [`for_file()`]. Then, change any of the decoder settings.
/// Finally, decode the input data with [`decode`].
/// ```
/// # use vpk0::{Encoder, Decoder};
/// let original = b"ABBACABBACD";
/// let compressed = Encoder::for_bytes(original)
///     .encode_to_vec()
///     .unwrap();
/// let decompressed = Decoder::for_bytes(&compressed)
///     .decode()
///     .unwrap();
/// assert_eq!(&original[..], decompressed);
/// ```
/// You can use a `Decoder` to get the [`VpkHeader`] with [`header()`]
///  or [`TreeInfo`] with [`trees()`]:
/// ```
/// # use vpk0::{Encoder, Decoder};
/// # let original = b"ABBACABBACD";
/// # let compressed = Encoder::for_bytes(original).encode_to_vec().unwrap();
/// let mut decoder = Decoder::for_bytes(&compressed);
/// let size = decoder.header().unwrap().size as usize;
/// assert_eq!(size, original.len());
/// ```
/// [`for_reader()`]: Decoder::for_reader
/// [`for_bytes()`]: Decoder::for_bytes
/// [`for_file()`]: Decoder::for_file
/// [`decode()`]: Decoder::decode
/// [`header()`]: Decoder::header
/// [`trees()`]: Decoder::trees
pub struct Decoder<'a, R: Read> {
    src: BitReader<R, BigEndian>,
    log: Option<LogWtr<'a>>,
    info: Option<(VpkHeader, RawTrees)>,
}

impl<'a, R: Read> Decoder<'a, R> {
    #[inline]
    pub fn for_reader(rdr: R) -> Self {
        Self {
            src: BitReader::endian(rdr, BigEndian),
            log: None,
            info: None,
        }
    }

    #[inline]
    pub fn with_logging<W: Write>(&mut self, wtr: &'a mut W) -> &mut Self {
        self.log = Some(wtr as LogWtr);
        self
    }

    #[inline]
    pub fn header(&mut self) -> Result<VpkHeader, VpkError> {
        self.get_file_info().map(|(hdr, _)| *hdr)
    }

    #[inline]
    pub fn trees(&mut self) -> Result<TreeInfo, VpkError> {
        self.get_file_info().map(|(_, t)| t.into())
    }

    #[inline]
    pub fn decode(&mut self) -> Result<Vec<u8>, VpkError> {
        do_decode(self)
    }

    fn get_file_info(&mut self) -> Result<&(VpkHeader, RawTrees), VpkError> {
        if let Some(ref info) = self.info {
            Ok(info)
        } else {
            let hdr = VpkHeader::from_bitreader(&mut self.src)?;
            let offsets = VpkTree::from_bitreader(&mut self.src)?;
            let lengths = VpkTree::from_bitreader(&mut self.src)?;

            self.info = Some((hdr, [offsets, lengths]));
            Ok(self.info.as_ref().unwrap())
        }
    }
}

impl<'a> Decoder<'a, Cursor<&'a [u8]>> {
    #[inline]
    pub fn for_bytes(bytes: &'a [u8]) -> Self {
        let rdr = Cursor::new(bytes);
        Self::for_reader(rdr)
    }
}

impl<'a> Decoder<'a, BufReader<File>> {
    #[inline]
    pub fn for_file<P: AsRef<Path>>(p: P) -> Result<Self, VpkError> {
        File::open(p)
            .map(BufReader::new)
            .map(Self::for_reader)
            .map_err(Into::into)
    }
}

/// Decompress a `Read`er of `vpk0` data into a `Vec<u8>`
///
/// This is a convenience function to decode a `Read`er without
/// having to import and set up a [`Decoder`]
pub fn decode<R: Read>(rdr: R) -> Result<Vec<u8>, VpkError> {
    Decoder::for_reader(rdr).decode()
}

/// Decompress a byte slice of `vpk0` data into a `Vec<u8>`
///
/// This is a convenience function to decode data without
/// having to import and set up a [`Decoder`] or a `std::io::Cursor`
pub fn decode_bytes(bytes: &[u8]) -> Result<Vec<u8>, VpkError> {
    Decoder::for_reader(Cursor::new(bytes)).decode()
}

/// Extract the [`VpkHeader`] and [`TreeInfo`] from `vpk0` data
///
/// This is a convenience function to extract information about `vpk0` data without having
/// to set up a [`Decoder`]
pub fn vpk_info<R: Read>(rdr: R) -> Result<(VpkHeader, TreeInfo), VpkError> {
    let mut decoder = Decoder::for_reader(rdr);

    decoder
        .header()
        .and_then(|hdr| decoder.trees().map(|t| (hdr, t)))
}

fn do_decode<R: Read>(opt: &mut Decoder<R>) -> Result<Vec<u8>, VpkError> {
    let info = if let Some(info) = opt.info.as_ref() {
        info
    } else {
        opt.get_file_info()?;
        opt.info.as_ref().unwrap()
    };
    let &(header, [ref offsets, ref lengths]) = info;
    let Decoder { src, log, .. } = opt;

    // set up the log with a map to store the bitsizes of the offsets and lengths
    let mut log = log.as_mut().map(|l| (l, LogFreq::new()));

    if let Some((wtr, _)) = &mut log {
        writeln!(wtr, "# Header\n{:?}", &header)?;
        writeln!(wtr, "## Offset / Moveback Tree\n{}", offsets)?;
        writeln!(wtr, "###> {:?}", offsets)?;
        writeln!(wtr, "## Length / Size Tree\n{}", lengths)?;
        writeln!(wtr, "###> {:?}", lengths)?;
        writeln!(wtr)?;
    }

    let output_size = header.size as usize;
    let mut output: Vec<u8> = Vec::with_capacity(output_size);

    while output.len() < output_size {
        if src.read_bit()? {
            let initial_move = offsets.read_value(src)? as usize;
            let move_back = match header.method {
                VpkMethod::TwoSample => {
                    if initial_move < 3 {
                        let l = initial_move + 1;
                        let u = offsets.read_value(src)? as usize;

                        if let Some((wtr, _)) = &mut log {
                            writeln!(
                                wtr,
                                "Encoded 2-sample => initial move: {} | second move: {}",
                                initial_move, u
                            )?;
                        }

                        (l + (u << 2)) - 8
                    } else {
                        if let Some((wtr, _)) = &mut log {
                            writeln!(wtr, "Encoded 2-sample => initial move: {}", initial_move)?;
                        }
                        (initial_move << 2) - 8
                    }
                }
                VpkMethod::OneSample => initial_move,
            };

            // get start position in output, and the number of bytes to copy-back
            if move_back > output.len() {
                return Err(VpkError::BadLookBack(move_back, output.len()));
            }

            let start = output.len() - move_back;
            let size = lengths.read_value(src)? as usize;

            if let Some((wtr, map)) = &mut log {
                let size_bits = usize::MAX.count_ones() - size.leading_zeros();
                let mb_bits = usize::MAX.count_ones() - move_back.leading_zeros();
                writeln!(
                    wtr,
                    "{:04x} - Encoded [Copyback]: size: {} ({} bits) mb: {} ({} bits) | start: {:04x}",
                    output.len(),
                    size,
                    size_bits,
                    move_back,
                    mb_bits,
                    start
                )?;
                *map.size.entry(size_bits as u8).or_insert(0) += 1;
                *map.moveback.entry(mb_bits as u8).or_insert(0) += 1;
            }

            for i in start..start + size {
                let byte = output[i];
                output.push(byte);
            }
            if let Some((wtr, _)) = &mut log {
                writeln!(wtr, "\t{:02x?}", &output[start..start + size])?;
            }
        } else {
            let byte = src.read(8)?;
            output.push(byte);

            if let Some((wtr, _)) = &mut log {
                writeln!(wtr, "{:04x} - Uncoded: {:02x}", output.len() - 1, byte)?;
            }
        }
    }

    Ok(output)
}

#[derive(Debug)]
struct LogFreq {
    size: BTreeMap<u8, u32>,
    moveback: BTreeMap<u8, u32>,
}

impl LogFreq {
    fn new() -> Self {
        Self {
            size: BTreeMap::new(),
            moveback: BTreeMap::new(),
        }
    }
}


src/errors.rs
use std::{
    error::Error,
    fmt, io,
    num::{ParseIntError, TryFromIntError},
    str,
};

/// Possible errors that arise from compressing or decompressing a `vpk0` binary
#[derive(Debug)]
#[non_exhaustive]
pub enum VpkError {
    InvalidHeader(String),
    InvalidMethod(u8),
    BadLookBack(usize, usize),
    BadTreeEncoding,
    BadUserTree(EncodeTreeParseErr),
    InputTooBig(TryFromIntError),
    Utf8Error(str::Utf8Error),
    Io(io::Error),
}

impl fmt::Display for VpkError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            VpkError::InvalidHeader(s) => write!(f, "Invalid ascii string '{}' in header", s),
            VpkError::InvalidMethod(n) => {
                write!(f, "VPK method {} is invalid and not supported", n)
            }
            VpkError::BadLookBack(mb, size) => write!(
                f,
                "Bad input file: asked to move back {} bytes in buffer of only {} bytes",
                mb, size
            ),
            VpkError::BadTreeEncoding => write!(f, "Huffman tree value couldn't be read"),
            VpkError::BadUserTree(_) => {
                write!(f, "Issue parsing user-provided huffman code tree string")
            }
            VpkError::InputTooBig(_) => write!(f, "Input file size too big to fit in 32-bit word"),
            VpkError::Utf8Error(_) => write!(f, "Couldn't read magic bytes"),
            VpkError::Io(_) => write!(f, "IO issue"),
        }
    }
}

impl Error for VpkError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            VpkError::BadUserTree(e) => Some(e as &dyn Error),
            VpkError::InputTooBig(e) => Some(e as &dyn Error),
            VpkError::Utf8Error(e) => Some(e as &dyn Error),
            VpkError::Io(e) => Some(e as &dyn Error),
            _ => None,
        }
    }
}

impl From<EncodeTreeParseErr> for VpkError {
    fn from(e: EncodeTreeParseErr) -> Self {
        Self::BadUserTree(e)
    }
}

impl From<io::Error> for VpkError {
    fn from(e: io::Error) -> Self {
        Self::Io(e)
    }
}

impl From<TryFromIntError> for VpkError {
    fn from(e: TryFromIntError) -> Self {
        Self::InputTooBig(e)
    }
}

/// Possible errors from parsing a user provided Huffman Tree
///
/// These errors can occur when a user passes a bad Huffman Tree to
/// the [`with_offsets`], [`with_lengths`], [`optional_offsets`], or [`optional_lengths`]
/// methods of a [`Encoder`](crate::Encoder).
///
/// [`with_offsets`]: crate::Encoder::with_offsets
/// [`with_lengths`]: crate::Encoder::with_lengths
/// [`optional_offsets`]: crate::Encoder::optional_offsets
/// [`optional_lengths`]: crate::Encoder::optional_lengths
#[derive(Debug)]
#[non_exhaustive]
pub enum EncodeTreeParseErr {
    LexNum(ParseIntError, usize),
    LexUnexp(char, usize),
    ParseUnexp(&'static str, usize),
    ParseUnexpEnd,
}

impl fmt::Display for EncodeTreeParseErr {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            EncodeTreeParseErr::LexNum(_, p) => {
                write!(f, "Issue parsing number in tree string at pos {}", p)
            }
            EncodeTreeParseErr::LexUnexp(c, p) => {
                write!(f, "Unexpected character '{}' at pos {}", c, p)
            }
            EncodeTreeParseErr::ParseUnexp(s, p) => {
                write!(f, "Unexpected token '{}' at pos {}", s, p)
            }
            EncodeTreeParseErr::ParseUnexpEnd => write!(f, "Unexpected end of tokens"),
        }
    }
}

impl Error for EncodeTreeParseErr {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            EncodeTreeParseErr::LexNum(e, _) => Some(e as &dyn Error),
            _ => None,
        }
    }
}


src/lib.rs
//! A Rust library for handling Nintendo's N64-era `vpk0` data compression
//!
//! `vpk0` is data compression scheme built on LZSS with Huffman coding for the
//! dictionary offsets and match lengths.
//! It was used at HAL Laboratory for three of their N64 games and later,
//! for encoding the data on GBA E-Reader cards.
//!
//! This crate provides decoding and encoding for `vpk0` files.
//! As it is an old scheme, the decoding and encoding is designed to match what
//! Nintendo did in the 1990s; the crate does not focus on compression ratio or speed.
//!
//! ## Usage
//! The [`decode()`] and [`encode()`] functions provide quick ways to deal with
//! `vpk0` data.
//!
//! ```
//! use vpk0::{encode, decode};
//! use std::io::Cursor;
//!
//! # let data = &b"AAAVVVAAABABACCCDE"[..];
//! let raw = Cursor::new(data);
//! let compressed = encode(raw).unwrap();
//! let decompressed = decode(Cursor::new(&compressed)).unwrap();
//! assert_eq!(&data, &decompressed);
//! ```
//!
//! For more control, you can use [`Decoder`] or [`Encoder`]:
//!
//! ```
//! use vpk0::Encoder;
//!
//! Encoder::for_bytes(b"ababacdcdeaba")
//!     .two_sample()
//!     .encode_to_writer(std::io::stdout())
//!     .unwrap();
//! ```
//! --------------------------------------------------------------------------------
//! ## `vpk0` Background
//! The `vpk0` format—named for the magic bytes—is thought to have been developed
//! at [HAL Laboratories] in the late 1990s. Three of their N64 games use the compression scheme:
//! * Super Smash Bros.
//! * Pokémon Snap
//! * Shigesato Itoi's No. 1 Bass Fishing: Definitive Edition
//!
//! The format next appeared in the mid-2000s as the compression used in the [Nintendo e-Reader]
//! for the GBA. This is where the format first received attention from the internet at large.
//! Tim Schuerewegen’s [nevpk] and Caitsith’s [NVPK Tool and NEDEC Make] were
//! open source implementations of `vpk0` that came from reverse engineering the e-Reader.
//!
//! This crate extends on their work to provide matching compression for HAL’s N64 titles.
//!
//! ## Format Overview
//! `vpk0` is based on two fundamental encoding algorithms: LSZZ and Huffman Coding.
//! The two techniques were bouncing around [the Japanese BSSes since the late 80s],
//! and together they comprise the backbone of many modern day encoding schemes like [Deflate].
//!
//! The `vpk0` format is comparatively simpler: it is a variable length LZSS.
//! The input data is compressed by a standard LZSS implementation.
//! But instead of having fixed bit sizes for the dictionary offset and length, the sizes are variable.
//! The variable bit sizes are then encoded as a Huffman code, with the necessary
//! [Huffman tree](format::TreeInfo) prepended to the encoded data.
//!
//! For more info, see [the documentation for the format module](crate::format)
//!
//! ## Implementation Details
//! This implementation is designed to be a byte-perfect match of the encoder used
//! for *Super Smash Bros.* As of March 2021, the LZSS encoder is byte-matching,
//! but the Huffman compression is not.
//!
//! The matching LZSS encoding scheme is: after a found match, look ahead at the next byte
//! to see if there is a longer match. Continue checking the next byte until
//! a smaller or no match is found.
//!
//! The encoder in this crate checks at most the next ten bytes,
//! as that was the maximum number necessary to match all 500 `vpk0` encoded files in *SSB64*.
//! In the future, this parameter may become another option for [`Encoder`].
//!
//! ## Advanced Usages
//! ### Getting info from a `vpk0` file
//! ```
//! use vpk0::vpk_info;
//! # use vpk0::Encoder;
//! # let vpkfile = Encoder::for_bytes(b"abababababacaaa").encode_to_vec().unwrap();
//! # let vpkfile = std::io::Cursor::new(vpkfile);
//! let (header, trees) = vpk_info(vpkfile).unwrap();
//! println!("Original size: {} bytes", header.size);
//! println!("VPK encoded with method {}", header.method);
//! println!("Offsets: {} || Lengths: {}", trees.offsets, trees.lengths);
//! ```
//!
//! ### Encode like a standard LZSS
//! ```
//! use vpk0::{Encoder, LzssSettings};
//! // use fixed length compression by setting the offset to 10 and the length to 6.
//! let compressed = Encoder::for_bytes(b"I am Sam. Sam I am.")
//!     .one_sample()
//!     .with_lzss_settings(LzssSettings::new(10, 6, 2))
//!     .with_offsets("10")
//!     .with_lengths("6")
//!     .encode_to_vec();
//! ```
//!
//! [HAL Laboratories]: https://www.hallab.co.jp/eng/
//! [Nintendo e-Reader]: https://en.m.wikipedia.org/wiki/Nintendo_e-Reader
//! [nevpk]: http://users.skynet.be/firefly/gba/e-reader/tools/index.htm
//! [NVPK Tool and NEDEC Make]: https://caitsith2.com/ereader/devtools.htm
//! [the Japanese BSSes since the late 80s]: https://web.archive.org/web/20160110174426/https://oku.edu.mie-u.ac.jp/~okumura/compression/history.html
//! [Deflate]: https://en.m.wikipedia.org/wiki/Deflate

mod decode;
mod encode;
pub mod errors;
pub mod format;

pub use decode::{decode, decode_bytes, vpk_info, Decoder};
pub use encode::{encode, encode_bytes, lzss::LzssSettings, Encoder, LzssBackend};
